inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager_genai_unit_test.py

```diff
@@ -0,0 +1,368 @@
+sys.path.insert(0, '/home/runner/GITHUB_ACTION_RUNNERS/_work/intel-inb-manageability/intel-inb-manageability/inbm/dispatcher-agent/')
+import pytest
+from unittest.mock import patch, MagicMock
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, ApplicationAddSourceParameters
+from unittest.mock import patch, mock_open
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, ApplicationSourceList
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, ApplicationRemoveSourceParameters, SourceError, LINUX_GPG_KEY_PATH, UBUNTU_APT_SOURCES_LIST_D
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, SourceError, ApplicationUpdateSourceParameters
+from dispatcher.source.ubuntu_source_manager import UbuntuOsSourceManager, SourceParameters, SourceError
+from dispatcher.source.ubuntu_source_manager import UbuntuOsSourceManager, SourceError
+from unittest.mock import patch, mock_open, call
+
+#DO NOT DELETE THIS LINE - TestUbuntuApplicationSourceManagerAdd
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuApplicationSourceManagerAdd:
+    @pytest.fixture
+    def mock_dispatcher_broker(self):
+        return MagicMock()
+
+    @pytest.fixture
+    def mock_application_source_manager(self, mock_dispatcher_broker):
+        return UbuntuApplicationSourceManager(mock_dispatcher_broker)
+
+    @pytest.mark.parametrize('gpg_key_name, gpg_key_uri, source_list_file_name, sources', [
+        ('key1', 'http://example.com/key1', 'file1', ['deb http://example.com/ubuntu/ bionic main restricted']),
+        (None, None, 'file2', ['deb http://example.com/ubuntu/ bionic main restricted']),
+    ])
+    @patch('dispatcher.source.ubuntu_source_manager.verify_source')
+    @patch('dispatcher.source.ubuntu_source_manager.add_gpg_key')
+    @patch('dispatcher.source.ubuntu_source_manager.create_file_with_contents')
+    def test_add(self, mock_create_file_with_contents, mock_add_gpg_key, mock_verify_source, mock_application_source_manager, gpg_key_name, gpg_key_uri, source_list_file_name, sources):
+        # Arrange
+        parameters = ApplicationAddSourceParameters(gpg_key_name=gpg_key_name, gpg_key_uri=gpg_key_uri, source_list_file_name=source_list_file_name, sources=sources)
+
+        # Act
+        mock_application_source_manager.add(parameters)
+
+        # Assert
+        if gpg_key_name and gpg_key_uri:
+            mock_verify_source.assert_called_once_with(source=gpg_key_uri.rsplit('/', 1)[0], dispatcher_broker=mock_application_source_manager._dispatcher_broker)
+            mock_add_gpg_key.assert_called_once_with(gpg_key_uri, gpg_key_name)
+        mock_create_file_with_contents.assert_called_once_with(f'/etc/apt/sources.list.d/{source_list_file_name}', sources)
+
+#DO NOT DELETE THIS LINE - TestUbuntuApplicationSourceManagerList
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuApplicationSourceManagerList:
```

Comment:
To go back to Pull Request Summary table, Click on the [link](https://github.com/intel/intel-inb-manageability/pull/467#)

<h3>Class: UbuntuApplicationSourceManager |Function: list</h3>

[Source Code Link](https://github.com/intel/intel-inb-manageability/blob/NEX_MAXIMUS_UT/develop/inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager.py?plain=1#L125)

[Unit Test Case Code Link](https://github.com/intel/intel-inb-manageability/blob/NEX_MAXIMUS_UT/develop/inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager_genai_unit_test.py?plain=1#L55)

<h4>Test Case Scenarios:</h4>
The unit test `test_list` in the `TestUbuntuApplicationSourceManagerList` class generates five test cases based on the parameters provided to the `pytest.mark.parametrize` decorator. Here are the test cases along with their explanations:

1. Test Case Name: test_list_with_one_source
   Explanation: This test case tests the `list` method with a file that contains one source. The expected output is a list with an `ApplicationSourceList` object that has a `sources` list with one source.

2. Test Case Name: test_list_with_multiple_sources
   Explanation: This test case tests the `list` method with a file that contains multiple sources. The expected output is a list with an `ApplicationSourceList` object that has a `sources` list with multiple sources.

3. Test Case Name: test_list_with_no_sources
   Explanation: This test case tests the `list` method with a file that does not contain any sources. The expected output is a list with an `ApplicationSourceList` object that has an empty `sources` list.

4. Test Case Name: test_list_with_commented_lines
   Explanation: This test case tests the `list` method with a file that contains commented lines. The `list` method should ignore these lines. The expected output is a list with an `ApplicationSourceList` object that has an empty `sources` list.

5. Test Case Name: test_list_with_blank_lines
   Explanation: This test case tests the `list` method with a file that contains blank lines. The `list` method should ignore these lines. The expected output is a list with an `ApplicationSourceList` object that has an empty `sources` list.

Each test case is designed to test a specific scenario that the `list` method may encounter when reading from a file. The test cases cover a wide range of possible inputs, including files with one source, multiple sources, no sources, commented lines, and blank lines.


<h4>Test Execution Results:</h4>

| Total Cases    | 5  |
| -------------- | -- |
| Passed Cases   | 5  |
| Failed Cases   | 0  |



To go back to Pull Request Summary table, Click on the [link](https://github.com/intel/intel-inb-manageability/pull/467#)